### PR TITLE
fix synchronization behavior for copies with type change

### LINKED ISSUE #121320: h2d non-blocking copy with type conversion doesn't respect stream semantics
I would expect `copy_` with `non_blocking=True` to follow stream semantics, if I have a few of them I'd expect them to run in order. However, this is not the case for h2d copy with type conversion:
```.py
import torch

FIFTY_MIL_CYCLES = 50000000

a=torch.ones(1, device="cuda")*10
b=torch.empty(1, device="cpu", pin_memory=True)
c=torch.empty(1, device="cuda", dtype=torch.long)

torch.cuda._sleep(FIFTY_MIL_CYCLES)
b.copy_(a, non_blocking=True)
c.copy_(b, non_blocking=True)
print(c) # prints 0, not 10
# without type conversion it works
c_float = torch.empty(1, device="cuda", dtype=torch.float)
b.fill_(0)
torch.cuda._sleep(FIFTY_MIL_CYCLES)
b.copy_(a, non_blocking=True)
c_float.copy_(b, non_blocking=True)
print(c_float) # prints 10.
```
currently we do conversion on the gpu for cpu-to-gpu copies (fix to #79933), there were some concerns at the time that using extra gpu memory could lead to ooms, but I think we should eat memory overhead to not have surprising behavior like above. 

cc @ptrblck

### PR DESCRIPTION
**Pull request by ngimel**
Fixes #121320

This is BC-breaking for non-blocking copies between CPU <-> GPU (either way) that also changes the dtype. Instead of doing the dtype conversion on the CPU, we now do it on the GPU. This will increase the GPU memory usage as a temporary buffer needs to be created on the GPU.

### CODE REVIEW

File: aten/src/ATen/native/cuda/Copy.cu [hunk 1]
```diff
@@ -302,9 +302,11 @@ static void copy_kernel_cuda(TensorIterator& iter, bool non_blocking) {
     Tensor src_contig;
 
     // If non_blocking is true - type conversions are performed on the GPU
-    // for CPU-GPU copies, otherwise type conversions are performed on the CPU.
```
Comment (albanD) [inline aten/src/ATen/native/cuda/Copy.cu:305]: I'm confused by this comment. For GPU -> CPU copies with non_blocking, the old code was already doing to dtype conversion on the src device which is the GPU, not the CPU?

Comment (ngimel) [reply aten/src/ATen/native/cuda/Copy.cu:305]: Yeah the comment was wrong (or they have a different meaning for `CPU-GPU copy`, after all it doesn't specify if it's `CPU-to-GPU` or `to-CPU-from-GPU`. Implied prepositions are important!

**Review by albanD — APPROVED** [review]
SGTM!

**Pull request closed without merging**